lib/compat/wordpress-6.0/class-gutenberg-rest-global-styles-controller.php

```diff
@@ -0,0 +1,96 @@
+<?php
+/**
+ * REST API: Gutenberg_REST_Global_Styles_Controller class
+ *
+ * @package    Gutenberg
+ * @subpackage REST_API
+ */
+
+/**
+ * Base Global Styles REST API Controller.
+ */
+class Gutenberg_REST_Global_Styles_Controller extends WP_REST_Global_Styles_Controller {
+	/**
+	 * Registers the controllers routes.
+	 *
+	 * @return void
+	 */
+	public function register_routes() {
+		// List themes global styles.
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/themes/(?P<stylesheet>[\/\s%\w\.\(\)\[\]\@_\-]+)/variations',
+			array(
+				array(
+					'methods'             => WP_REST_Server::READABLE,
+					'callback'            => array( $this, 'get_theme_items' ),
+					'permission_callback' => array( $this, 'get_theme_items_permissions_check' ),
+					'args'                => array(
+						'stylesheet' => array(
+							'description' => __( 'The theme identifier', 'gutenberg' ),
+							'type'        => 'string',
+						),
+					),
+				),
+			)
+		);
+
+		parent::register_routes();
+	}
+
+	/**
+	 * Checks if a given request has access to read a single theme global styles config.
+	 *
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return true|WP_Error True if the request has read access for the item, WP_Error object otherwise.
+	 */
+	public function get_theme_items_permissions_check( $request ) {
+		// Verify if the current user has edit_theme_options capability.
+		// This capability is required to edit/view/delete templates.
+		if ( ! current_user_can( 'edit_theme_options' ) ) {
+			return new WP_Error(
+				'rest_cannot_manage_global_styles',
+				__( 'Sorry, you are not allowed to access the global styles on this site.', 'gutenberg' ),
+				array(
+					'status' => rest_authorization_required_code(),
+				)
+			);
+		}
+
+		return true;
+	}
+
+	/**
+	 * Returns the given theme global styles variations.
+	 *
+	 * @param WP_REST_Request $request The request instance.
+	 *
+	 * @return WP_REST_Response|WP_Error
+	 */
+	public function get_theme_items( $request ) {
+		if ( wp_get_theme()->get_stylesheet() !== $request['stylesheet'] ) {
+			// This endpoint only supports the active theme for now.
+			return new WP_Error(
+				'rest_theme_not_found',
+				__( 'Theme not found.', 'gutenberg' ),
+				array( 'status' => 404 )
+			);
+		}
+
+		$variations     = array();
+		$base_directory = get_stylesheet_directory() . '/styles';
+		if ( file_exists( $base_directory ) ) {
+			$nested_files      = new RecursiveIteratorIterator( new RecursiveDirectoryIterator( $base_directory ) );
+			$nested_html_files = new RegexIterator( $nested_files, '/^.+\.json$/i', RecursiveRegexIterator::GET_MATCH );
+			foreach ( $nested_html_files as $path => $file ) {
+				$decoded_file = wp_json_file_decode( $path, array( 'associative' => true ) );
+				if ( is_array( $decoded_file ) ) {
+					$variations[] = $decoded_file;
```

Comment:
This returns the theme.json unprocessed (so no migrations of versions, no validation, no normalization). This is because the current APIs are very tied to the active theme and the theme.json file.

I think we need a `load_theme_json` file or something like that to reuse here and in the theme.json class. cc @oandregal (would also be useful to process inactive themes)

Reply:
All those things will be done by using `new WP_Theme_JSON( $decoded_file )->get_raw_data()`.

However, there's a couple of other things that raise some questions for me:

- Does variations need to be translated (preset names)? I guess this boils down to: can variations provide things other than styles (settings, presets)?
- How do we know at runtime which variation is selected by the user to generate the global stylesheet in the frontend?
- What happens with child themes? Do they show the parent variations in addition to their own?
- How data is sent to the site editor changed a bit during the 5.9 so I don't remember off the top of my head the full state of it. By skimming through this class it seems we're sending only the theme data (variations) via this endpoint. In the site editor, we'll need to merge the theme data with the default data. I presume there's a different endpoint for that as well or should we do the merge here?

Reply:
If variations can add settings I'm not sure we can keep querying for `get_theme_support` to backfill things into theme.json: the theme support API only works with the active theme.

Reply:
> Does variations need to be translated (preset names)? I guess this boils down to: can variations provide things other than styles (settings, presets)?

Yes, I think they can

> How do we know at runtime which variation is selected by the user to generate the global stylesheet in the frontend?

We don't care about this information, the variation will be saved as "user changes" overriding the default theme.json

> What happens with child themes? Do they show the parent variations in addition to their own?

For now, I decided that they don't but we can change that at any time

> we'll need to merge the theme data with the default data. I presume there's a different endpoint for that as well or should we do the merge here?

we don't need to merge anything, since the variations are strictly equivalent to the "user" part (user changes), so they'll just work similarly, we only send the "user changes" right now for the user changes.

Reply:
> If variations can add settings I'm not sure we can keep querying for get_theme_support to backfill things into theme.json: the theme support API only works with the active theme.

I'm not sure how this impacts my work here but for me, theme supports are just legacy default theme.json properties (and not allowed in variations)

Reply:
> we don't need to merge anything, since the variations are strictly equivalent to the "user" part (user changes), so they'll just work similarly, we only send the "user changes" right now for the user changes.

Ah, nice. So the "default" variation serves as a reset to the user preferences? Meaning that it isn't stored as user data because it's already taken into account as theme data?

> re: variations providing things other than styles

If we also need translations (presets, custom templates, etc) I can work that out later/in parallel to this PR. Probably the `translate` operation could be moved to within the `WP_Theme_JSON` class as another operation similar to the migrations, normalization, etc.

Reply:
> Ah, nice. So the "default" variation serves as a reset to the user preferences? Meaning that it isn't stored as user data because it's already taken into account as theme data?

yes, that's how it works now, though I didn't add the default to the list in the UI yet.

> If we also need translations (presets, custom templates, etc) I can work that out later/in parallel to this PR. Probably the translate operation could be moved to within the WP_Theme_JSON class as another operation similar to the migrations, normalization, etc.

👍